cxx/attribute.hh

```diff
@@ -0,0 +1,128 @@
+    // to/unincorporate from clusters themselves rather than relations. Empty
+    // clusters are not deleted, because observations will be re-incorporated to
+    // them. The number of observations in each cluster is the same before/after
+    // the call to this method.
+    base_relation->unincorporate_from_cluster(base_items);
+
+    // Unincorporate all noisy observations from clusters and store the noisy
+    // values.
+    std::unordered_map<std::string,
+                       std::unordered_map<T_items, ValueType, H_items>>
+        noisy_observations;
+    std::vector<ValueType> all_noisy_observations;
+    for (auto [name, rel] : noisy_relations) {
+      noisy_observations[name] = {};
+      for (const T_items& items : rel->base_to_noisy_items.at(base_items)) {
+        const ValueType& v = rel->get_value(items);
+        rel->unincorporate_from_cluster(items);
+        noisy_observations[name][items] = v;
+        all_noisy_observations.push_back(v);
+      }
+    }
+
+    // Compute logp_score of the noisy relations without the observations of
+    // this attribute.
+    double baseline_logp_score = 0.;
+    for (auto [name, rel] : noisy_relations) {
+      baseline_logp_score += rel->logp_score();
+    }
+
+    // Candidates for the next "true" value are the clean proposals from each
+    // cluster in each noisy relation.
+    std::vector<ValueType> true_value_candidates;
+    for (auto [name, rel] : noisy_relations) {
+      for (const auto& [i, em] : rel->emission_relation.clusters) {
+        true_value_candidates.push_back(
+            reinterpret_cast<Emission<ValueType>*>(em)->propose_clean(
+                all_noisy_observations, prng));
+      }
+    }
+
+    std::vector<double> logpx;
+    for (const ValueType& v : true_value_candidates) {
+      // Incorporate the true/noisy values for this candidate.
+      base_relation->incorporate_to_cluster(base_items, v);
+      for (auto& [name, rel] : noisy_relations) {
+        for (const T_items& items : rel->base_to_noisy_items.at(base_items)) {
+          rel->incorporate_to_cluster(items,
+                                      noisy_observations.at(name).at(items));
+        }
+      }
+
+      // Compute the incremental logp_score of the noisy relations.
+      double candidate_logp_score = 0.;
+      for (const auto& [name, rel] : noisy_relations) {
+        candidate_logp_score += rel->logp_score();
+      }
+      logpx.push_back(candidate_logp_score - baseline_logp_score);
+
+      // Unincorporate the true/noisy values for this candidate.
+      for (auto& [name, rel] : noisy_relations) {
+        for (const T_items& items : rel->base_to_noisy_items.at(base_items)) {
+          rel->unincorporate_from_cluster(items);
+        }
+      }
+      base_relation->unincorporate_from_cluster(base_items);
+    }
+    
+    // Sample a new latent value.
+    ValueType new_true_value = true_value_candidates[log_choice(logpx, prng)];
```

Comment:
If we want this to be the maximum instead of a sample, that would simplify the code substantially -- `incorporate_to_cluster` and `unincorporate_from_cluster` could be replaced with a single `replace_in_cluster` method (since we wouldn't need `baseline_logp_score`), which would be safer since it would never leave the relation in an invalid state. If we want the option of sampling instead of taking the maximum then I think we have to do something like what's here.